src/nvim/tui/input.c

```diff
@@ -282,27 +289,56 @@ static bool handle_focus_event(TermInput *input)
 
 static bool handle_bracketed_paste(TermInput *input)
 {
-  if (rbuffer_size(input->read_stream.buffer) > 5 &&
-      (!rbuffer_cmp(input->read_stream.buffer, "\x1b[200~", 6) ||
-       !rbuffer_cmp(input->read_stream.buffer, "\x1b[201~", 6))) {
-    bool enable = *rbuffer_get(input->read_stream.buffer, 4) == '0';
-    // Advance past the sequence
-    rbuffer_consumed(input->read_stream.buffer, 6);
-    if (input->paste_enabled == enable) {
-      return true;
+  RBuffer *rbuf = input->read_stream.buffer;
+  if (rbuffer_size(rbuf) > 5
+      && (!rbuffer_cmp(rbuf, "\x1b[200~", 6)
+          || !rbuffer_cmp(rbuf, "\x1b[201~", 6))) {
+    bool enable = *rbuffer_get(rbuf, 4) == '0';
+
+    if (enable && input->paste_started) {
+      // Received a bogus "paste start" after paste was already started.
+      return false;
+    }
+
+    rbuffer_consumed(rbuf, 6);  // Advance past the sequence
+    input->paste_started = enable;
+    if (enable) {
+      loop_schedule(&loop, event_create(1, apply_pastepre, 0));
+    } else {
+      enqueue_input(input, PASTEPOST_KEY, sizeof(PASTEPOST_KEY) - 1);
```

Comment:
> @tarruda So does that mean we can safely use loop_schedule() to schedule FocusGained/FocusLost instead of the K_FOCUSGAINED/K_FOCUSLOST?

Possibly yes. I may be missing something, but FocusGained/FocusLost doesn't seem to be events that needs to be handled at an exact point during input processing.